File: stack/0020-valid-parentheses/java/Solution.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public boolean isValid(String s) {
        Deque<Character> stack = new ArrayDeque<>();

        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else {
                if (stack.isEmpty()) return false;

                char top = stack.pop();
                boolean matches = (c == ')' && top == '(') ||
                                   (c == '}' && top == '{') ||
                                   (c == ']' && top == '[');
                if (!matches) return false;
            }
        }

        return stack.isEmpty();
    }
}

File: stack/0020-valid-parentheses/cpp/Solution.cpp
#include <stack>
#include <string>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;

        for (char c : s) {
            if (c == '(' || c == '{' || c == '[') {
                st.push(c);
            } else {
                if (st.empty()) return false;

                char top = st.top();
                st.pop();
                bool matches = (c == ')' && top == '(') ||
                               (c == '}' && top == '{') ||
                               (c == ']' && top == '[');
                if (!matches) return false;
            }
        }

        return st.empty();
    }
};
